File: DSA/Heap/Code/MinHeap.java
// Min Heap in Java

import java.util.*;

public class MinHeap {
    private List<Integer> arr;
    
    public MinHeap() {
        this.arr = new ArrayList<>();
    }
    
    private int parent(int i) {
        return (i - 1) / 2;
    }
    
    private int leftChild(int i) {
        return 2 * i + 1;
    }
    
    private int rightChild(int i) {
        return 2 * i + 2;
    }
    
    private void swap(int i, int j) {
        int temp = arr.get(i);
        arr.set(i, arr.get(j));
        arr.set(j, temp);
    }
    
    private void heapifyUp(int i) {
        while (i > 0 && arr.get(parent(i)) > arr.get(i)) {
            swap(i, parent(i));
            i = parent(i);
        }
    }
    
    private void heapifyDown(int i) {
        int smallest = i;
        int left = leftChild(i);
        int right = rightChild(i);
        
        if (left < arr.size() && arr.get(left) < arr.get(smallest)) {
            smallest = left;
        }
        if (right < arr.size() && arr.get(right) < arr.get(smallest)) {
            smallest = right;
        }
        
        if (smallest != i) {
            swap(i, smallest);
            heapifyDown(smallest);
        }
    }
    
    public void insert(int key) {
        arr.add(key);
        heapifyUp(arr.size() - 1);
    }
    
    public int extractMin() {
        if (arr.isEmpty()) return -1;
        
        int min = arr.get(0);
        arr.set(0, arr.get(arr.size() - 1));
        arr.remove(arr.size() - 1);
        
        if (!arr.isEmpty()) {
            heapifyDown(0);
        }
        
        return min;
    }
    
    public int getMin() {
        return arr.isEmpty() ? -1 : arr.get(0);
    }
    
    public void printHeap() {
        System.out.println("Min Heap (Array): " + arr);
    }
    
    public static void main(String[] args) {
        System.out.println("=== Min Heap (Java) ===\n");
        
        MinHeap heap = new MinHeap();
        
        System.out.println("Inserting: 10, 5, 3, 2, 8, 1");
        heap.insert(10);
        heap.insert(5);
        heap.insert(3);
        heap.insert(2);
        heap.insert(8);
        heap.insert(1);
        
        heap.printHeap();
        
        System.out.println("\nExtracting minimum elements:");
        System.out.println("Extract: " + heap.extractMin());
        heap.printHeap();
        System.out.println("Extract: " + heap.extractMin());
        heap.printHeap();
        
        System.out.println("\n=== Min Heap Properties ===");
        System.out.println("1. Parent <= Left Child");
        System.out.println("2. Parent <= Right Child");
        System.out.println("3. Complete binary tree (all levels filled except last)");
        System.out.println("4. Minimum element always at root\n");
        
        System.out.println("=== Min Heap Complexity ===");
        System.out.println("Insert:      O(log n)");
        System.out.println("Extract Min: O(log n)");
        System.out.println("Get Min:     O(1)");
        System.out.println("Space:       O(n)");
    }
}

File: DSA/Heap/Code/MinHeap.py
# Min Heap in Python

class MinHeap:
    def __init__(self):
        self.arr = []
    
    def parent(self, i):
        return (i - 1) // 2
    
    def left_child(self, i):
        return 2 * i + 1
    
    def right_child(self, i):
        return 2 * i + 2
    
    def swap(self, i, j):
        self.arr[i], self.arr[j] = self.arr[j], self.arr[i]
    
    def heapify_up(self, i):
        while i > 0 and self.arr[self.parent(i)] > self.arr[i]:
            self.swap(i, self.parent(i))
            i = self.parent(i)
    
    def heapify_down(self, i):
        smallest = i
        left = self.left_child(i)
        right = self.right_child(i)
        
        if left < len(self.arr) and self.arr[left] < self.arr[smallest]:
            smallest = left
        if right < len(self.arr) and self.arr[right] < self.arr[smallest]:
            smallest = right
        
        if smallest != i:
            self.swap(i, smallest)
            self.heapify_down(smallest)
    
    def insert(self, key):
        self.arr.append(key)
        self.heapify_up(len(self.arr) - 1)
    
    def extract_min(self):
        if not self.arr:
            return -1
        
        min_val = self.arr[0]
        self.arr[0] = self.arr[-1]
        self.arr.pop()
        
        if self.arr:
            self.heapify_down(0)
        
        return min_val
    
    def get_min(self):
        return self.arr[0] if self.arr else -1
    
    def print_heap(self):
        print("Min Heap (Array):", self.arr)

# Test
print("=== Min Heap (Python) ===\n")

heap = MinHeap()

print("Inserting: 10, 5, 3, 2, 8, 1")
heap.insert(10)
heap.insert(5)
heap.insert(3)
heap.insert(2)
heap.insert(8)
heap.insert(1)

heap.print_heap()

print("\nExtracting minimum elements:")
print(f"Extract: {heap.extract_min()}")
heap.print_heap()
print(f"Extract: {heap.extract_min()}")
heap.print_heap()

print("\n=== Min Heap Properties ===")
print("1. Parent <= Left Child")
print("2. Parent <= Right Child")
print("3. Complete binary tree (all levels filled except last)")
print("4. Minimum element always at root\n")

print("=== Min Heap Complexity ===")
print("Insert:      O(log n)")
print("Extract Min: O(log n)")
print("Get Min:     O(1)")
print("Space:       O(n)")
